adapters/repos/db/vector/flat/index.go

```diff
@@ -462,28 +473,54 @@ func (index *flat) searchByVectorBQ(ctx context.Context, vector []float32, k int
 		idsSlice.slice[i] = heap.Pop().ID
 	}
 
+	// rescore
+
 	// we expect to be mostly IO-bound, so more goroutines than CPUs is fine
 	distancesUncompressedVectors := make([]float32, len(idsSlice.slice))
 
+	// use consistent setting for entire rescore – only get feature flag once
+	rescoreAgainstObjectStore := false
+	if index.rescoreAgainstObjectStore != nil {
+		rescoreAgainstObjectStore = index.rescoreAgainstObjectStore.Get()
+	}
+
 	eg := enterrors.NewErrorGroupWrapper(index.logger)
 	for workerID := 0; workerID < index.concurrentCacheReads; workerID++ {
 		workerID := workerID
 		eg.Go(func() error {
 			for idPos := workerID; idPos < len(idsSlice.slice); idPos += index.concurrentCacheReads {
 				id := idsSlice.slice[idPos]
-				candidateAsBytes, err := index.vectorById(id)
-				if err != nil {
-					return err
-				}
-				if len(candidateAsBytes) == 0 {
-					continue
-				}
-				distance, err := distanceCalc(candidateAsBytes)
-				if err != nil {
-					return err
-				}
+				if !rescoreAgainstObjectStore {
+					candidateAsBytes, err := index.vectorById(id)
+					if err != nil {
+						return fmt.Errorf("rescore: retrieve vector: %w", err)
+					}
+					if len(candidateAsBytes) == 0 {
+						continue
+					}
+					distance, err := distanceCalc(candidateAsBytes)
+					if err != nil {
+						return fmt.Errorf("rescore: distance calc: %w", err)
+					}
+
+					distancesUncompressedVectors[idPos] = distance
+				} else {
+					// this part is experimental, but may end up being better than the
+					// "standard" way, see
+					// usecases/config/runtime/flat_index_rescore_strategy.go for details
 
-				distancesUncompressedVectors[idPos] = distance
+					candidate, err := index.vectorForIDThunk(ctx, id)
+					if err != nil {
+						return fmt.Errorf("rescore (object store): retrieve vector: %w", err)
+					}
+
+					distance, err := index.distancerProvider.SingleDist(vector, candidate)
+					if err != nil {
+						return fmt.Errorf("rescore (object store): distance calc: %w", err)
+					}
```

Comment:
```suggestion
					distance, err := index.DistanceBetweenVectors(vector, candidate)
					if err != nil {
						return fmt.Errorf("rescore (object store): %w", err)
					}
```

Should this re-use DistanceBetweenVectors? (we can also cut redundancy from the error message)